src/plugins/vis_augmenter/public/vega/helpers.ts

```diff
@@ -0,0 +1,324 @@
+ */
+const convertToDateTimeObj = (timestamp: number): any => {
+  const momentObj = moment(timestamp);
+  return {
+    year: Number(momentObj.format('YYYY')),
+    month: momentObj.format('MMMM'),
+    date: momentObj.date(),
+    hours: momentObj.hours(),
+    minutes: momentObj.minutes(),
+    seconds: momentObj.seconds(),
+    milliseconds: momentObj.milliseconds(),
+  };
+};
+
+export const generateVisLayerFilterString = (visLayerColumnIds: string[]): string => {
+  if (!isEmpty(visLayerColumnIds)) {
+    const filterString = visLayerColumnIds.map(
+      (visLayerColumnId) => `datum['${visLayerColumnId}'] > 0`
+    );
+    return filterString.join(' || ');
+  } else {
+    // if there is no VisLayers to display, then filter out everything by always returning false
+    return 'false';
+  }
+};
+
+/**
+ * By default, the source datatable will not include rows with empty data.
+ * For handling events that may belong in missing buckets that are not yet
+ * created, we need to create them. For more details, see description in
+ * https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3145
+ *
+ * Note that this may add buckets with start/end times out of the chart bounds.
+ * This is the current default behavior of histogram aggregations with intervals,
+ * in order for the bucket keys to have "clean" timestamp keys (e.g., 1/1 @ 12AM).
+ * For more details, see
+ * https://opensearch.org/docs/latest/opensearch/bucket-agg/#histogram-date_histogram
+ *
+ * Also note this is only adding empty buckets at the beginning/end of a table. We are
+ * not taking into account missing buckets within source datapoints. Because of this
+ * limitation, it is possible that charted events may not be put into the most precise
+ * bucket based on their raw event timestamps, if there is missing / sparse source data.
+ */
+export const addMissingRowsToTableBounds = (
+  datatable: OpenSearchDashboardsDatatable,
+  dimensions: any
+): OpenSearchDashboardsDatatable => {
+  const augmentedTable = cloneDeep(datatable);
+  const intervalMillis = moment.duration(dimensions.x.params.interval).asMilliseconds();
+  const xAxisId = getXAxisId(dimensions, augmentedTable.columns);
+  const chartStartTime = new Date(dimensions.x.params.bounds.min).valueOf();
+  const chartEndTime = new Date(dimensions.x.params.bounds.max).valueOf();
+
+  if (!isEmpty(augmentedTable.rows)) {
+    const dataStartTime = augmentedTable.rows[0][xAxisId] as number;
+    const dataEndTime = augmentedTable.rows[augmentedTable.rows.length - 1][xAxisId] as number;
+
+    let curStartTime = dataStartTime;
+    while (curStartTime > chartStartTime) {
+      curStartTime -= intervalMillis;
+      augmentedTable.rows.unshift({
+        [xAxisId]: curStartTime,
+      });
+    }
+
+    let curEndTime = dataEndTime;
+    while (curEndTime < chartEndTime) {
+      curEndTime += intervalMillis;
+      augmentedTable.rows.push({
+        [xAxisId]: curEndTime,
+      });
```

Comment:
Please read the description above the fn. Basically the bounds of the datatable may be smaller than the actual bounds of the chart. So we are expanding the data rows to have consistent bounds with the actual chart.

Overall, we have to do this since we are removing the `domain` field from the x-axis config (which also was causing the issues with running the charts locally), and instead will be using the default bounds found from the source datatable.